src/v/storage/log_manager.cc

```diff
@@ -259,20 +244,132 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+             * we were sleeping then reschedule rather than run immediately.
+             * this attempts to avoid thundering herd since config changes are
+             * delivered immediately to all shards.
+             */
+            if (_jitter.base_duration() != prev_jitter_base) {
+                continue;
+            }
+        } catch (const ss::semaphore_timed_out&) {
+            // time for some chores
+        }
 
-    return ss::with_scheduling_group(
-      _config.compaction_sg, [this, collection_threshold] {
-          return housekeeping_scan(collection_threshold);
-      });
+        /*
+         * When we are in a low disk space situation we would like to reclaim
+         * data as fast as possible since being in that state may cause all
+         * sorts of problems like blocking producers, or even crashing the
+         * system. The fastest way to do this is to apply retention rules to
+         * partitions in the order of most to least amount of reclaimable data.
+         * This amount can be estimated using the log::disk_usage(gc_config)
+         * interface.
+         */
+        if (
+          _disk_space_alert == disk_space_alert::degraded
+          || _disk_space_alert == disk_space_alert::low_space) {
+            /*
+             * build a schedule of partitions to gc ordered by amount of
+             * estimated reclaimable space. since logs may be asynchronously
+             * deleted doing this safely is tricky.
+             */
+            absl::btree_map<size_t, model::ntp, std::greater<>> ntp_by_gc_size;
+
+            /*
+             * first we build a collection of ntp's as their estimated
+             * reclaimable disk space. the loop and disk_usage() are safe
+             * against concurrent log removals.
+             */
+            for (auto& log_meta : _logs_list) {
+                /*
+                 * applying segment.ms will make reclaimable data from the
+                 * active segment visible.
+                 */
+                co_await log_meta.handle.apply_segment_ms();
+                if (!log_meta.link.is_linked()) {
+                    continue;
+                }
+
+                auto ntp = log_meta.handle.config().ntp();
+                auto log = dynamic_cast<disk_log_impl*>(
+                  log_meta.handle.get_impl());
+                auto usage = co_await log->disk_usage(
+                  gc_config(collection_threshold(), _config.retention_bytes()));
+
+                /*
+                 * NOTE: this estimate is for local retention policy only. for a
+                 * policy that considers removing data uploaded to the cloud,
+                 * this estimate is available in `usage.reclaim.available`.
+                 */
+                ntp_by_gc_size.emplace(usage.reclaim.retention, std::move(ntp));
+            }
+
+            /*
+             * Since we don't hold on to a per-log gate after we've recorded it
```

Comment:
Would be nice to have some kind of shared ptr here, or if we could use the intrusive list links somehow, though maybe that's not possible.

Reply:
yeh. the concurrency requirements for the log object and the collection of logs in the manager is whack. i mean, it isn't just about using a shared_ptr--the log object already is--it's about the concurrent operations and log shutdown.